Installer: queue only link/run deps unless building from source

becker33 opened the pull request:
Depends on #39136 (commits up-to/including 39ff are from that PR)

- Refactors BuildTask into separate classes BuildTask and InstallTask
- Queues all packages as InstallTask, with link/run deps only
- If an InstallTask fails to install from binary, a BuildTask is generated
- The BuildTask is queued with dependencies on the new InstallTasks for its build deps and their link/run dependencies.
- The Tasks telescope open to include all build deps of build deps ad-hoc

TODO:
- [ ] Docs
- [x] Fix InstallStatus to track/report new dependencies properly

becker33 committed:
Installer: queue only link/run deps and requeue with build deps as needed

Refactors BuildTask into separate classes BuildTask and InstallTask
Queues all packages as InstallTask, with link/run deps only
If an InstallTask fails to install from binary, a BuildTask is generated
The BuildTask is queued with dependencies on the new InstallTasks for its
build deps and their link/run dependencies.
The Tasks telescope open to include all build deps of build deps ad-hoc

tldahlgren committed:
Resolve mypy issues

becker33 committed:
update edges for existing tasks for build deps

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -2466,7 +2707,11 @@ def install(self):
         try:
             with fs.replace_directory_transaction(self.task.pkg.prefix):
-                self.installer._install_task(self.task, self.install_status)
+                rc = self.installer._install_task(self.task, self.install_status)
```
🤔 "Reaching into" the installer's private method makes me wonder if there's a better design option here.

becker33 replied:
Not really related to this, but yes I agree -- I'll see if I can handle this at the same time.

tldahlgren reviewed (commented):
I'm still reviewing this but wanted to add the comment.

becker33 committed:
refactor overwrite installs into main installer class

becker33 committed:
PackageInstaller._install_task: fix type annotation

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1955,6 +1955,42 @@ def _install_task(self, task: Task, install_status: InstallStatus) -> None:
+                "The spec is now uninstalled."
+            )
+
+            # Unwrap the actuall installation exception
```
Typo: `actuall` -> `actual`.

becker33 committed:
rebase fixup

tldahlgren committed:
Installer: update installation progress tracking
- test/installer: use existing inst for spack.installer
- remove install status from Installing message
- Add specs count visitor
- Report status on installed plus minor refactor
- Add the "+" to the tracker; include one experimental dynamic calculation
- tweak status reporting to include ensuring numerator unique across installed packages
- _print_installed_pkg -> InstallStatus.print_installed()
- move set_term_title outside of InstallStatus
- InstallStatus: remove unnecessary next_pkg
- InstallStatus: class and method name changes
  * changed InstallStatus to InstallerStatus since already have former in
    database.py and spec.py
  * changed print_installed to set_installed since does more than print now
- InstallerStatus -> InstallerProgress, install_status -> progress
- InstallerProgress: cache config:install_status
- InstallerProgress: restore get_progress and set_term_title methods (w/ tweaks)
- Task execute(): added returns to docstrings
- Don't pass progress to build_process or Installer.run, but set installed on successful return
- fix mypy issue with pkg.run_tests assignment

becker33 committed:
fix issues after rebase

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
cleanup

Signed-off-by: Gregory Becker <becker33@llnl.gov>
